hips/hip-9999.md

```diff
@@ -0,0 +1,72 @@
+authors: [ "Ralf Van Dorsselaer" ]
+created: "2022-06-20"
+type: "feature"
+status: "draft"
+---
+ 
+## Abstract
+ 
+Helm release data encoded into the release field in a Kubernetes Secret can cause the Secret as a whole to exceed 1 MiB (as defined by MaxSecretSize in kubernetes/pkg/apis/core/types.go).
+ 
+When this happens, the Kubernetes Secret cannot be stored in Etcd and the Helm install or upgrade fails.
+ 
+Splitting the Helm release data across multiple Kubernetes Secrets resolves this problem.
+ 
+## Motivation
+ 
+Helm release data that causes the Kubernetes Secret size to exceed MaxSecretSize will make Helm install or upgrade fail.
+ 
+The existing HELM_DRIVER=secret storage driver does not handle Kubernetes Secrets become too big when storing a large encoded Helm release data.
+ 
+## Rationale
+ 
+The logical place to support this is in the HELM_DRIVER=secret storage driver.
+ 
+## Specification
+ 
+The feature is transparent to users. 
+ 
+When the HELM_DRIVER=secret storage driver notices that the resulting Kubernetes Secret will exceed MaxSecretSize then the storage driver will split the encoded release data across multiple Kubernetes Secrets.
+ 
+Splitting requires additional metadata to be added to the Kubernetes Secret:
+ 
+- number of chunks: integer, the number of Kubernetes Secrets that holds the release data for this Helm release version.
+- chunk: integer, the chunk number. Each Secret gets a chunk number. Reconstituting the release data can be done by reading the Secret with chunk=1, then subsequently reading in numerical order any number of additional Secrets up to and including the "number of chunks".
+ 
+Splitting requires the name of the Secret to include a "chunk" suffix. Example: sh.helm.release.v1.test.v1.2, 3, 4 and so on where the number after the last dot is the "number of chunk". The first Secret will not get a .1 suffix.
+ 
+If no splitting is required then no additional metadata is included/set.
+ 
+## Backwards compatibility
+ 
+If the chunking is implemented in the secret/secrets HELM_DRIVER then backward compatibility to previously installed Helm releases is guaranteed. If no chunking information is encoded in the release secret then the driver can behave as before chunking support was added. If no chunking is needed then the release secret can be created without any additional metadata and with compatible name.
+ 
+## Security implications
+ 
+Theoretically a MTM Man in The Middle attack is possible where requests to the Kubernetes API server are intercepted. But this type of attack applies to any access to the Kubernetes API server.
+ 
+## How to teach this
+ 
+The feature is transparent to users.
+ 
+The documentation (in advanced section(s)) needs to explain that multiple Kubernetes Secrets can be created for a Helm release to counter the size limitation.
+ 
+## Reference implementation
+ 
+PR: https://github.com/helm/helm/pull/11087
+ 
+## Rejected ideas
+ 
+n/a
```

Comment:
What about using the database driver? That is a known supported driver that works with releases larger than 1MB in size.

Reply:
In our case, we are a software vendor who provide helm-based applications for customers to deploy in their cluster, we are not the user of helm itself.  With that, we have to work with typical helm setups; asking for a certain minimum version of helm (e.g. for defect fixes in helm) has never been a problem and customers seemed to be current anyways, but asking to configured a PG only for helm would introduce a level of complexity that I'm not sure customers would accept, and I have not seen a single customer who has configured their helm with a PG storage backend.

However, if you're worried about changing the size-constrained etcd-storage backend, is it worth considering to introduce a parallel, chunked-etcd-storage backend that is in parallel?  That would avoid the customer complexity of having to deploy a Postgresql and allows to use the storage that is guaranteed to be available.  I believe the approach that @ralfv suggested with extending the existing etcd backend is more elegant and simpler, though, but it might address your 3.0.0 compatiblity concern.

Reply:
> What about using the database driver? That is a known supported driver that works with releases larger than 1MB in size.
> 
> https://helm.sh/docs/topics/advanced/#sql-storage-backend

A big part of Helm (certainly v3) is that it does not require any additional infrastructure besides a Kubernetes cluster. I really want to avoid introducing additional infrastructure to satisfy the requirement to store Secrets that are larger than 1MB.

Reply:
> might address your 3.0.0 compatiblity concern.

@fravoss I think the compatibility concern goes away if support for chunked secrets is implemented starting with Helm v4 and converting is transparently handled between Secrets < 1MB and Secrets > 1MB - as is curently done in the initial implementation see https://github.com/helm/helm/pull/11087 - as Helm v3 sill simply be unable to process such Secrets there is no harm. It would also be possible to backport to Helm v3 obviously.